tools/codegen/README.md

```diff
@@ -78,3 +74,100 @@ update-codegen-crds:
 
 You may also want to add the `_output` directory to your `.gitignore` to avoid checking in compiled binaries created
 by this make target.
+
+## Generators
+
+The following section describes the individual generators included within the `codegen` utility.
+
+The generators enabled by default are:
+- [Compatibility](#compatibility)
+- [Schemapatch](#schemapatch)
+- [Swaggerdocs](#swaggerdocs)
+
+### Compatibility
+
+To generate API compatibility comments, use the `compatibility` subcommand.
+
+The `compatibility` subcommand generates a compatibility level comment for each API defintion.
+The generation is controlled by a marker applied to the CRD struct defintiion.
+For example, this annotation would be `+openshift:compatibility-gen:level=1` for a level 1 API.
+	
+Valid API levels are 1, 2, 3 and 4. Version 1 is required for all stable APIs.
+Version 2 should be used for beta level APIs. Levels 3 and 4 may be used for alpha APIs.
+
+### Schemapatch
+
+To generate CRD schemas, use the `schemapatch` subcommand.
+
+The schemapatch subcommand uses the [controller-tools][controller-tools] project to generate CRD schemas.
+It requires a stub CRD to exist before the schema will be generated. Copying an existing OpenShift CRD
```

Comment:
another idea for future us, why bother with the stub.  An opt-out comment on a type containing typemeta and objectmeta seems more practical.

Reply:
I assume there is some historic reason for why we were using the schema patch instead of the full generator. I wonder if the full generation would clear the annotations that we set?

I could see a future version of this tool looking for a CRD yaml, if it doesn't exist, creating one, else, run the schema patch.

One irritation of the schemapatch tool is that it doesn't maintain customer printer columns, so that's another thing we can solve when we solve the stub issue